numpy/core/src/multiarray/mapping.c

```diff
@@ -1783,35 +1780,35 @@ array_assign_subscript(PyArrayObject *self, PyObject *ind, PyObject *op)
         return -1;
     }
 
-    /* Single field access */
-    if (PyDataType_HASFIELDS(PyArray_DESCR(self))) {
-        if (PyString_Check(ind) || PyUnicode_Check(ind)) {
-            PyObject *obj;
+    /* field access */
+    if (PyDataType_HASFIELDS(PyArray_DESCR(self)) &&
+            obj_is_string_or_stringlist(ind)) {
+        PyObject *obj;
+        static PyObject *indexfunc = NULL;
 
-            obj = PyDict_GetItem(PyArray_DESCR(self)->fields, ind);
-            if (obj != NULL) {
-                PyArray_Descr *descr;
-                int offset;
-                PyObject *title;
+        if (!(PyString_Check(ind) || PyUnicode_Check(ind))) {
```

Comment:
Numpy does define a `PyUString_Check`, but that does not check for unicode in python2. In descriptor.c I see a lot of

    #if defined(NPY_PY3K)
            if (PyUString_Check(title)) {
    #else
            if (PyUString_Check(title) || PyUnicode_Check(title)) {
    #endif

I don't see the point of using `PyUString` there since in the top it always compiles to `PyUnicode_Check(title)` and in the bottom always to `PyString_Check(title)`.

I'll update this to just use the usual checks, but maybe in the future it might be nice to define a macro to simplify all those cases.